alphazero/_game.py

```diff
@@ -383,3 +384,339 @@ def n_players():
+            tensor = torch.Tensor(board_state)
+            if width * height != tensor.numel():
+                raise ValueError(
+                    f"width and height dimensions {width}, {height} do not "
+                    f"match board state size {tensor.shape}"
+                )
+            self._board = tensor.view(height, width)
+        self._move_count = 0
+        self._players = [0, 1]
+        self.win_list = [False, False]
+
+    def make_move(self, move: int):
+        """
+        Make a move on the board. Should be a valid move and return a valid
+        game board.
+        """
+        self._board = self.board_after_move(
+            self._board, self._players[self._move_count % 2], move
+        )
+        self._move_count += 1
+
+        if not self.is_valid():
+            raise ValueError(f"Board state {self._board} is not valid.")
+
+    @staticmethod
+    def board_after_move(board, player: int, move: int):
+        """
+        Make a move on the board and return it. Does not affect any instances
+        of the class. Should be a valid move and return a valid game board.
+        """
+        if not isinstance(move, int):
+            raise TypeError(f"move is of type {type(move)} but must be an int")
+
+        if player not in (0, 1):
+            raise ValueError(f"invalid player for move {move}")
+
+        if any((board[:, move] == -1)) is False:
+            raise ValueError(f"Column is full")
+        if move > board.shape[0]:
+            raise IndexError(
+                f"invalid move {move} for board shaped {board.shape}"
+            )
+
+        row_index = board.shape[0] - 1
+        while row_index >= 0 and board[row_index, move] != -1:
+            row_index -= 1
+
+        new_board = board.clone()
+        new_board[row_index, move] = player
+
+        return new_board
+
+    @staticmethod
+    def get_initial_board(height=6, width=7):
+        """
+        Get the initial game board for this game. For Connect 4, for instance,
+        this should be an empty board of width `width` and height `height`, but
+        for chess this may be the initial position of the pieces.
+        """
+        return torch.full((height, width), -1)
+
+    @staticmethod
+    def get_legal_moves(board) -> torch.Tensor:
+        """
+        Get a list of legal moves for the given board game position. Should
+        return a `torch.Tensor` full of booleans that represent whether a move
+        is valid or not. Does not affect internal state of any instances of
+        this class.
+        """
+        return board[0, :] == -1
+
+    def current_legal_moves(self) -> torch.Tensor:
+        """
+        Get a dynamic list of legal moves for the current game position. Should
+        return a `torch.Tensor` full of booleans that represent whether a move
+        is valid or not.
+        """
+        return self.get_legal_moves(self._board)
+
+    @staticmethod
+    def winning_move(board, col) -> bool:
```

Comment:
i'm not going to look at this function lol. make sure to put in some tests (if you haven't already) to make sure that this thing is bulletproof please :)

Reply:
So after troubleshooting for a while, I can't really get the function to work. I tried
```python
board = torch.Tensor(
            [
                [-1, -1, -1, -1, -1, -1, -1],
                [-1, -1, -1, -1, -1, -1, -1],
                [-1, -1, -1, -1, -1, -1, -1],
                [-1, -1, -1, -1, -1, -1, -1],
                [-1, -1, -1, -1, -1, -1, -1],
                [0, -1, -1, -1, -1, -1, -1],
            ]
        )
Connect4.winning_move(board, 1)
```
It should return False, but I kept getting True(Also tried winning_move(board, 1)) @homerours Not sure if it worked for you? There was an extra 
```
row += 1
```
that was making the code break, but after getting rid of it I thought the problem would be solved... As for now we'll use get_game_status, and I have removed winning_move, but I have saved it locally in case we need to revisit later.